### ISSUE TITLE
node_namespace_pod_container:container_cpu_usage_seconds_total:sum_rate

### ISSUE THREAD

**Issue by xiaozhangzhang1** [prometheus-operator/kube-prometheus#1056]
```
record: node_namespace_pod_container:container_cpu_usage_seconds_total:sum_rate
expr: sum by(cluster, namespace, pod, container) (rate(container_cpu_usage_seconds_total{container!="POD",image!="",job="kubelet",metrics_path="/metrics/cadvisor"}[5m])) * on(cluster, namespace, pod) group_left(node) topk by(cluster, namespace, pod) (1, max by(cluster, namespace, pod, node) (kube_pod_info{node!=""}))
```

this record rule is not work in promethues ,if i change on(cluster, namespace, pod) is on( namespace, pod),it works  

* Prometheus Operator version:
      release-0.6

* Kubernetes version information:

    `kubectl version`
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.14", GitCommit:"89182bdd065fbcaffefec691908a739d161efc03", GitTreeState:"clean", BuildDate:"2020-12-18T12:11:25Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.14", GitCommit:"89182bdd065fbcaffefec691908a739d161efc03", GitTreeState:"clean", BuildDate:"2020-12-18T12:02:35Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}

*

**Comment by paulfantom** [comment]
It looks like you have a `cluster` label in one metric but not in the other. Does `container_cpu_usage_seconds_total{container!="POD",image!="",job="kubelet",metrics_path="/metrics/cadvisor", cluster!=""}` and `kube_pod_info{node!="",cluster!=""}` return any output?

**Comment by xiaozhangzhang1** [comment]
> It looks like you have a `cluster` label in one metric but not in the other. Does `container_cpu_usage_seconds_total{container!="POD",image!="",job="kubelet",metrics_path="/metrics/cadvisor", cluster!=""}` and `kube_pod_info{node!="",cluster!=""}` return any output?

i did, return   no data

**Comment by xiaozhangzhang1** [comment]
> It looks like you have a `cluster` label in one metric but not in the other. Does `container_cpu_usage_seconds_total{container!="POD",image!="",job="kubelet",metrics_path="/metrics/cadvisor", cluster!=""}` and `kube_pod_info{node!="",cluster!=""}` return any output?

kube_pod_info{node!="",cluster!=""}

**Comment by xiaozhangzhang1** [comment]
> It looks like you have a `cluster` label in one metric but not in the other. Does `container_cpu_usage_seconds_total{container!="POD",image!="",job="kubelet",metrics_path="/metrics/cadvisor", cluster!=""}` and `kube_pod_info{node!="",cluster!=""}` return any output?

kube_pod_info{node!="",cluster!=""}   return many

**Comment by xiaozhangzhang1** [comment]
container_cpu_usage_seconds_total{container!="POD",image!="",job="kubelet",metrics_path="/metrics/cadvisor", cluster!=""}
return no data

**Comment by paulfantom** [comment]
It seems that your cluster is not configured correctly and you have `cluster` label attached to metrics from kube-state-metrics, but not to metrics from kubelet. You need to have it in both places.

**Comment by xiaozhangzhang1** [comment]
> It seems that your cluster is not configured correctly and you have `cluster` label attached to metrics from kube-state-metrics, but not to metrics from kubelet. You need to have it in both places.

thanks ,i got it ,yes ,i did  cluster label to kube-state-metrics, i did  the same to kubelet, but  it not works ,

**Comment by ArchiFleKs** [comment]
I have the same issue where CPU usage is not working anymore on grafana

**Comment by Andor** [comment]
It is broken with kube-state-metrics 2.0 and newer: https://github.com/kubernetes/kube-state-metrics/issues/1501
There is a workaround, but it is not super easy to apply with kube-prometheus.

**Comment by rouja** [comment]
Hi,

I'm not sure but maybe this issue came from this change : 

https://github.com/prometheus-operator/kube-prometheus/commit/78a467737064513e64eeb2b1df60a2478cfeb23c

It seems that the record _node_namespace_pod_container:container_cpu_usage_seconds_total:sum_rate_ was replaced by _node_namespace_pod_container:container_cpu_usage_seconds_total:sum__**i**_rate_

**Comment by SonalJain1707** [comment]
I am also facing same issue

sum(node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate{cluster="sandbox", namespace="admin"}) / sum(kube_pod_container_resource_requests{job="kube-state-metrics", cluster="sandbox", namespace="admin", resource="cpu"})

Does not return data

**Comment by rmn-lux** [comment]
+1 the same thing 

> I am also facing same issue
> 
> sum(node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate{cluster="$cluster", namespace="$namespace"}) / sum(kube_pod_container_resource_requests{job="kube-state-metrics", cluster="$cluster", namespace="$namespace", resource="cpu"})
> 
> Does not return data

**Comment by jeremydescamps** [comment]
When querying `node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate` on my prometheus instance, that does not return anything.

**Comment by fguiet** [comment]
Hi there,

Prometheus stack chart : kube-prometheus-stack-43.1.1, App Version: 0.61.1
K8s deployed with Rancher Docker version 2.7 : 1.24.4

To me, it was related to this issue : <https://github.com/k3s-io/k3s/issues/5782>
As mentioned in the issue, image label is now missing.

Workaournd : I removed the `image!=""` llabel in all the rules from `prometheus-stack-kube-prom-k8s.rules.yaml` file and now my grafana dashboard work like a charm

```bash
# Extract from file : prometheus-stack-kube-prom-k8s.rules.yaml
- name: k8s.rules
      rules:
        - expr: >-
            sum by (cluster, namespace, pod, container) (
              irate(container_cpu_usage_seconds_total{job="kubelet", metrics_path="/metrics/cadvisor", image!=""}[5m])
            ) * on (cluster, namespace, pod) group_left(node) topk by (cluster,
            namespace, pod) (
              1, max by(cluster, namespace, pod, node) (kube_pod_info{node!=""})
            )
          record: >-
            node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate
```

![image](https://user-images.githubusercontent.com/13849639/209106940-ab683915-0750-48b3-a14b-854fa83a3cc8.png)

**Comment by bmgante** [comment]
Hi @fguiet 
Got stuck in this problem as well, i am using latest helm chart version. 
I've already removed the label `image!=""` from k8s.rules and cpu dashboards started working. 

However, i still have issues for memory dashboard which basically use metric `container_memory_working_set_bytes`.

Example:
`sum(container_memory_working_set_bytes{job="kubelet", metrics_path="/metrics/cadvisor", cluster="$cluster", namespace="$namespace", pod="$pod", container!="", image!=""}) by (container)`

From prometheus, there is no cluster, container or image labels for this metric. Did you face this issue as well and if yes, how did you fix it?

<img width="1497" alt="image" src="https://user-images.githubusercontent.com/27358817/225988552-db01558d-e380-4a45-b67f-938d2014d9ba.png">

Thanks

**Comment by sirajkrm** [comment]
FWIW this has been addressed in later versions of Rancher Server 2.6.11 along upgrading k8s to `1.24.10-rancher4-1`

**Comment by anthosz** [comment]
> When querying `node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate` on my prometheus instance, that does not return anything. From what exporter this metric come from ?

Same for me, did you found it finally?

Using the last helm chart of prometheus stack

**Comment by zuchka** [comment]
I'm hitting this as well. any ideas?

**Comment by DhruvPatel2647** [comment]
if you have included  this  in the values of prometheus :

before: kubelet:
   serviceMonitor:
     https: false
     
 After (This works): 
 kubelet:
   serviceMonitor:
     https: true 
  because Kubelet is responsible for that metrics.
  
  for me I have disabled http in service-monitor for kubelt then I research it and foud that kublelt hhtp shoulbe enabled that is http:true

**Comment by gustavofbreunig** [comment]
I'm facing this same issue, reported [here](https://github.com/rancher/charts/issues/2271). 

Removed container!="" and image!="" from prometheus-stack-kube-prom-k8s.rules.yaml worked.

**Comment by mohamadkhani** [comment]
Thanks to @gustavofbreunig.
Removing all `image!=""` from  `charts/kube-prometheus-stack/templates/prometheus/rules-1.14/k8s.rules.yaml` file, fixed my problem too.

This is [my fork](https://github.com/mohamadkhani/helm-charts.git) if any one wants to check it.

**Comment by gustavofbreunig** [comment]
It was a rancher issue, corrected on `v1.24.10-rancher4-1`

https://github.com/rancher/rancher/issues/38934

proceed to close the issue

**Comment by nathanmcgarvey-modopayments** [comment]
> Related issue: [google/cadvisor#3336](https://github.com/google/cadvisor/issues/3336)

Also related details if you are on Docker-Desktop: https://github.com/docker/for-mac/issues/6969
